engine/gctscript.go

```diff
@@ -0,0 +1,82 @@
+package engine
+
+import (
+	"fmt"
+	"path/filepath"
+	"sync/atomic"
+
+	"github.com/thrasher-corp/gocryptotrader/gctscript/vm"
+	log "github.com/thrasher-corp/gocryptotrader/logger"
+)
+
+const gctscriptManagerName = "GCTScript"
+
+type gctScriptManager struct {
+	running  atomic.Value
+	shutdown chan struct{}
+}
+
+// Started returns if gctscript manager subsystem is started
+func (g *gctScriptManager) Started() bool {
+	return g.running.Load() == true
+}
+
+// Start starts gctscript subsystem and creates shutdown channel
+func (g *gctScriptManager) Start() error {
+	if g.Started() {
+		return fmt.Errorf("%s %s", gctscriptManagerName, ErrSubSystemAlreadyStarted)
+	}
+	log.Debugln(log.Global, gctscriptManagerName, MsgSubSystemStarting)
+	g.shutdown = make(chan struct{})
+	go g.run()
+	return nil
+}
+
+// Stop stops gctscript subsystem along with all running Virtual Machines
+func (g *gctScriptManager) Stop() error {
+	if !g.Started() {
+		return fmt.Errorf("%s %s", gctscriptManagerName, ErrSubSystemAlreadyStopped)
+	}
+
+	log.Debugln(log.Global, gctscriptManagerName, MsgSubSystemShuttingDown)
+	close(g.shutdown)
```

Comment:
I was in the midst of converting all of our managers to using atomic.Value but figured that we'd introduce a very small chance race condition and potential crash in case someone (an app using this) calls Stop twice before the manager _does_ actually shutdown (close on an already closed channel). The original way stops this with the usage of 2 vars (started/stopped) to know if it's started and in the process of stopping and will throw an error up until the slate is set back to 0/0 for started and stopped

Reply:
Just pushed a commit that reverts back to this style of start/stop management

Reply:
👍